src/nupic/algorithms/SDRClassifier.cpp

```diff
@@ -209,33 +227,44 @@ vector<Real64> SDRClassifier::calculateError_(const vector<UInt> &bucketIdxList,
   // compute predicted likelihoods
   vector<Real64> likelihoods(maxBucketIdx_ + 1, 0);
 
-  for (auto &bit : patternNZ) {
-    const Matrix &weights = weightMatrix_.at(step);
-    add(likelihoods.begin(), likelihoods.end(), weights.begin(bit),
-        weights.begin(bit + 1));
+  for (const auto& bit : patternNZ) {
+    auto& rowWeights = weightMatrix_.at(step).at(bit); //row
+    for(Size i=0; i< likelihoods.size(); i++) {
+      likelihoods[i] += rowWeights[i];
+    }
   }
   softmax_(likelihoods.begin(), likelihoods.end());
 
   // compute target likelihoods
   vector<Real64> targetDistribution(maxBucketIdx_ + 1, 0.0);
-  Real64 numCategories = (Real64)bucketIdxList.size();
+  const Real64 numCategories = (Real64)bucketIdxList.size();
   for (size_t i = 0; i < bucketIdxList.size(); i++)
     targetDistribution[bucketIdxList[i]] = 1.0 / numCategories;
-
-  axby(-1.0, likelihoods, 1.0, targetDistribution);
+  
+  NTA_ASSERT(likelihoods.size() == targetDistribution.size());
+  for(UInt i = 0; i < likelihoods.size(); i++) {
+    likelihoods[i] = targetDistribution[i] - likelihoods[i];
+  }
   return likelihoods;
 }
 
-void SDRClassifier::softmax_(vector<Real64>::iterator begin,
-                             vector<Real64>::iterator end) {
-  vector<Real64>::iterator maxItr = max_element(begin, end);
-  for (auto itr = begin; itr != end; ++itr) {
-    *itr -= *maxItr;
+
+void SDRClassifier::softmax_(const vector<Real64>::iterator begin,
```

Comment:
@htm-community/nupic-cpp I'd like your help reviewing this `softmax_` function. I think I found an error in upstream implementation. It is not a big deal (didn't even have a test to spot it, but let's make sure it's right)

- I'm reworking SDRClassifier, and its softmax_()
- wrote test SDRClassifierTest::testSoftmax to compare results (with results obtained from master) -> and we diff! 
  - added debug statements
  - backported test and debug to master, you can view it at temporary PR # 

- the "original" softmax_ version from master uses: 
  - range_exp()  from math/ArrayAlgo.hpp
    - Exp<double>() from math/Math.hpp 
  - accumulate() - I'm not sure if std::, or something from SparseMatrix::accumulate

Debug results from master: 
```
max=2.018
in=0
in=1
in=1.337
in=2.018
in=1.1
in=0.5
in=0.9
exp=0.132921
exp=0.361317
exp=0.506111
exp=1
exp=3.00417
exp=1.64872
exp=2.4596
sum=9.11284
div=0.0145861
div=0.0396492
div=0.0555382
div=0.109735
div=0.329663
div=0.180923
div=0.269905
[       OK ] SDRClassifierTest.testSoftmax (1 ms)
[----------] 1 test from SDRClassifierTest (1 ms total)
``` 

and from this PR:
```
max=2.018
in=0
exp=0.132921
in=1
exp=0.361317
in=1.337
exp=0.506111
in=2.018
exp=1
in=1.1
exp=0.399317
in=0.5
exp=0.21915
in=0.9
exp=0.326933
sum=2.94575
div=0.045123
div=0.122657
div=0.171811
div=0.339472
div=0.135557
div=0.0743953
div=0.110985
[  FAILED  ] SDRClassifierTest.testSoftmax

 1 FAILED TEST
```

You can see a strange behavior in the old/master implementation after "2.018 aka max elemen aka where exp=1": 
the old function gives exp > 1, while our does return < 1 (as it should). Because we normalized the values to the max element( x[i]-maxVal) , all values should be <= 1. 

TL;DR: 
 - not too critical (no test even before)
 - I think ours new is correct, just want to doublecheck

Reply:
I went ahead with our version